File: Leetcode/322. Coin Change/322_Coin_Change.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        vector<int> dp(amount + 1, amount + 1);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++)
        {
            for (int c = 0; c < coins.size(); c++)
            {
                if (i - coins[c] >= 0)
                    dp[i] = min(dp[i], dp[i - coins[c]] + 1);
            }
        }
        if (dp[amount] == amount + 1)
            return -1;

        return dp[amount];
    }
};

int main()
{
    vector<int> coins = {1, 2, 5};
    Solution s;
    int amount = 11;
    cout << s.coinChange(coins, amount) << endl;
}

File: Leetcode/322. Coin Change/322_Coin_Change.py
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0

        for i in range(amount + 1):
            for c in coins:
                if i - c >= 0:
                    dp[i] = min(dp[i], 1 + dp[i - c])

        return dp[amount] if dp[amount] != amount + 1 else -1

class Main:
    def main(self):
        coins = [1, 2, 5]
        amount = 11

        solution = Solution()
        print(solution.coinChange(coins, amount))
    
if __name__ == "__main__":
    Main().main()